### DISCUSSION TITLE
Making Upgrades Much Much Smoother

### DISCUSSION THREAD

**Discussion by discoverdefiteam** [terpnetwork/terp-core#2] [Ideas]
Upgrade v0.1.1 will be used as an opportunity to improve the overall upgrade experience for contributors. 

In order to do so, we should include:

- Any necessary architecture to make upgrade life easier
- Ready and reliable documentation for cosmovisor upgrades
- Workflow documentation for upgrading testing process (multi-node testing scripts for both hard upgrade and cosmovisor).

Will use this discussion to organize any useful references 👍

**Comment by discoverdefiteam** [comment]
Beginning with some previous issue referenced shared, lets see if theres anything we are missing from this info:
## Issues
~ [Issue #8538](https://github.com/cosmos/cosmos-sdk/issues/8538) (cosmos-sdk)
        Takeaways: 
        -  using `--unsafe-skip-upgrade`, The nodes then came up and starting participating in consensus, but never came to make a new block, even though we have > 67% voting power in pre-commits.
        - Akash network testnet had similar issue, reverted back to the original chain afterwards.
        - [Issue #111](https://github.com/e-money/em-ledger/issues/111): Remove and or replace support for time based upgrade #111 (e-Money)
~ [Issue #10188](https://github.com/cosmos/cosmos-sdk/issues/10188) (cosmos-sdk)

## Pulls
~ forced deterministic iteration order in x/upgrade and store during upgrades [Pull #10189](https://github.com/cosmos/cosmos-sdk/pull/10189) 

## Relevant Files
 `x/upgrade/keeper/keeper.go`: https://github.com/tomtau/cosmos-sdk/blob/741712167e8268f15c1c0edf046ba5ff3bc6377f/x/upgrade/keeper/keeper.go

**Comment by discoverdefiteam** [comment]
Another closely relevant improvement that can be included is referenced here: [fix: panic on export zero-height because variable-length addresses #1015](https://github.com/cosmos/gaia/pull/1015)

**Comment by hard-nett** [comment]
See [#1139](https://github.com/CosmWasm/wasmd/issues/1139)

**Comment by hard-nett** [comment]
This tutorial can be used to follow the upgrade processes: https://medium.com/web3-surfers/cosmos-dev-series-cosmos-sdk-based-blockchain-upgrade-b5e99181554c
